airflow/metrics/otel_logger.py

```diff
@@ -0,0 +1,289 @@
+        :param count: A positive integer to add to the current value of stat.
+        :param rate: value between 0 and 1 that represents the sampled rate at
+            which the metric is going to be emitted.
+        :param tags: Tags to append to the stat.
+        """
+        if (count < 0) or (rate < 0):
+            raise ValueError("count and rate must both be positive values.")
+        if rate < 1 and random.random() > rate:
+            return
+
+        if self.metrics_validator.test(stat) and name_is_otel_safe(self.prefix, stat):
+            counter = self.metrics_map.get_counter(f"{self.prefix}.{stat}", attributes=tags)
+            counter.add(count, attributes=tags)
+            return counter
+
+    def decr(
+        self,
+        stat: str,
+        count: int = 1,
+        rate: float = 1,
+        tags: Attributes = None,
+    ):
+        """
+        Decrement stat by count.
+
+        :param stat: The name of the stat to decrement.
+        :param count: A positive integer to subtract from current value of stat.
+        :param rate: value between 0 and 1 that represents the sampled rate at
+            which the metric is going to be emitted.
+        :param tags: Tags to append to the stat.
+        """
+        if (count < 0) or (rate < 0):
+            raise ValueError("count and rate must both be positive values.")
+        if rate < 1 and random.random() > rate:
+            return
+
+        if self.metrics_validator.test(stat) and name_is_otel_safe(self.prefix, stat):
+            counter = self.metrics_map.get_counter(f"{self.prefix}.{stat}")
+            counter.add(-count, attributes=tags)
+            return counter
+
+    @validate_stat
+    def gauge(
+        self,
+        stat: str,
+        value: int | float,
+        rate: float = 1,
+        delta: bool = False,
+        *,
+        tags: Attributes = None,
+    ) -> None:
+        warnings.warn("OpenTelemetry Gauges are not yet implemented.")
+        return None
+
+    @validate_stat
+    def timing(
+        self,
+        stat: str,
+        dt: DeltaType,
+        *,
+        tags: Attributes = None,
+    ) -> None:
+        warnings.warn("OpenTelemetry Timers are not yet implemented.")
+        return None
+
+    @validate_stat
+    def timer(
+        self,
+        stat: str | None = None,
+        *args,
+        tags: Attributes = None,
+        **kwargs,
+    ) -> TimerProtocol:
+        warnings.warn("OpenTelemetry Timers are not yet implemented.")
+        return Timer()
+
+
+class MetricsMap:
+    """Stores Otel Instruments."""
+
+    def __init__(self, meter):
+        self.meter = meter
+        self.map = {}
+
+    def clear(self) -> None:
+        self.map.clear()
+
+    def _create_counter(self, name):
+        """Creates a new counter or up_down_counter for the provided name."""
+        otel_safe_name = name[:OTEL_NAME_MAX_LENGTH]
+        if name != otel_safe_name:
+            warnings.warn(
```

Comment:
Should we simply raise error in this case? The problem with this warning is that user who will see it will not be able to do anything with it, and those are the worst kinds of warnings. This check should be targeted not for the users of airflow, but for those who create new metrics, and for them, just failing it here would be IMHO much betters because it will prevent them from adding the wrong metrics name in the first place.

Reply:
Until/unless we deprecate all of the old names that are too long, I think this needs to stay.  I can add a TODO comment noting that it can be changed to an exception if the old names are removed, but there are existing names which would start breaking if we made this an exception.

Reply:
I agree, it would be awesome to strictly prohibit names that break OTEL. I think a middle ground can be struck here:
maintain a list of the existing metric names that are too long, and warn for those, anything else/new throw an exception. Then we can work to deprecate the "grandfathered in" long metric names overtime.

Reply:
We did discuss this at the beginning of his project in the terms of a "metrics registry". The problem with having a list of names which are too long is that most (all?) of the names which would cause an issue are dynamically generated.  

Some are easy to catch with a "startswith()" like  [f"local_task_job.task_exit.{self.job.id}.{self.task_instance.dag_id}.{self.task_instance.task_id}.{return_code}"](https://github.com/apache/airflow/blob/main/airflow/jobs/local_task_job_runner.py#L298), but then we would also have [f"dagrun.{dag.dag_id}.first_task_scheduling_delay"](https://github.com/apache/airflow/blob/main/airflow/models/dagrun.py#L929) and [<job_name>_heartbeat_failure](https://github.com/apache/airflow/blob/main/airflow/jobs/job.py#L215) to figure out how to match.  

I guess we could use some regex matching shenanigans, if we want t do this.

This also goes back to the early discussions:  Is OTel going to be one available option for metrics alongside StatsD, or is it going to be the only option?   If it is going to be a pluggable option as we decided, then I don't feel like enforcing its naming policies universally is a great idea.

Reply:
> Some are easy to catch with a "startswith()" like f"local_task_job.task_exit.{self.job.id}.{self.task_instance.dag_id}.{self.task_instance.task_id}.{return_code}", but then we would also have f"dagrun.{dag.dag_id}.first_task_scheduling_delay" and <job_name>_heartbeat_failure to figure out how to match.
>
> I guess we could use some regex matching shenanigans, if we want t do this.

Yeah I think a regex pattern that you register for each one would be a nice solution.

> If it is going to be a pluggable option as we decided, then I don't feel like enforcing its naming policies universally is a great idea.

I don't think it's unreasonable to enforce the minimum name length across all possible metrics platforms?

Reply:
I've added a `set` which has all of the currently potentially-too-long names as regex patterns and added a check against that.  If the name is too long and not on that exemption list then an exception is thrown.  Let me know hat you two thin of the way I implemented it.  This should let us deprecate the old names over time and prevent anyone from adding new metrics that would break the standard.